### PR TITLE
 BUG: Nrows cannot be zero for read_csv. Fixes #21141

### LINKED ISSUE #21141: read_csv errors when low_memory=True, index_col is not None, and nrows=0
#### Code Sample

```python
import pandas as pd
from io import StringIO

data = """A,B,C,D,E
2000-01-03 00:00:00,0.980268513777,3.68573087906,-0.364216805298,-1.15973806169,foo
2000-01-04 00:00:00,1.04791624281,-0.0412318367011,-0.16181208307,0.212549316967,bar
2000-01-05 00:00:00,0.498580885705,0.731167677815,-0.537677223318,1.34627041952,baz
2000-01-06 00:00:00,1.12020151869,1.56762092543,0.00364077397681,0.67525259227,qux
2000-01-07 00:00:00,-0.487094399463,0.571454623474,-1.6116394093,0.103468562917,foo2"""

df = pd.read_csv(StringIO(data), low_memory=True, index_col=0, nrows=0)
```
#### Problem description

The above code results in `TypeError: 'NoneType' object is not iterable`. `read_csv` behaves correctly if `low_memory=False`, `index_col=None` or `nrows>0`.

Traceback:
```
Traceback (most recent call last):
  File "/home/peter/workspace/ray_env/lib/python3.6/site-packages/pandas/io/parsers.py", line 1848, in read
    data = self._reader.read(nrows)
  File "pandas/_libs/parsers.pyx", line 876, in pandas._libs.parsers.TextReader.read
  File "pandas/_libs/parsers.pyx", line 916, in pandas._libs.parsers.TextReader._read_low_memory
StopIteration

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "pd_bug.py", line 12, in <module>
    df = pd.read_csv(StringIO(data), low_memory=True, index_col=0, nrows=0)
  File "/home/peter/workspace/ray_env/lib/python3.6/site-packages/pandas/io/parsers.py", line 678, in parser_f
    return _read(filepath_or_buffer, kwds)
  File "/home/peter/workspace/ray_env/lib/python3.6/site-packages/pandas/io/parsers.py", line 446, in _read
    data = parser.read(nrows)
  File "/home/peter/workspace/ray_env/lib/python3.6/site-packages/pandas/io/parsers.py", line 1036, in read
    ret = self._engine.read(nrows)
  File "/home/peter/workspace/ray_env/lib/python3.6/site-packages/pandas/io/parsers.py", line 1855, in read
    dtype=self.kwds.get('dtype'))
  File "/home/peter/workspace/ray_env/lib/python3.6/site-packages/pandas/io/parsers.py", line 3215, in _get_empty_meta
    data = [Series([], dtype=dtype[name]) for name in index_names]
TypeError: 'NoneType' object is not iterable
```

#### Expected Output
No error.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.3.final.0
python-bits: 64
OS: Linux
OS-release: 4.13.0-36-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.23.0
pytest: 3.3.1
pip: 9.0.1
setuptools: 38.4.0
Cython: 0.27.3
numpy: 1.14.3
scipy: 1.0.0
pyarrow: 0.8.0
xarray: None
IPython: 6.2.1
sphinx: None
patsy: None
dateutil: 2.7.3
pytz: 2018.4
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: 0.4.0
matplotlib: 2.1.1
openpyxl: 2.5.3
xlrd: 1.1.0
xlwt: None
xlsxwriter: None
lxml: 4.2.0
bs4: 4.6.0
html5lib: 0.999999999
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by cgopalan**
- [x] closes #21141
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Comment by gfyoung** [comment]
@jreback : Do we need to deprecate `nrow=0` before disallowing?  It's a (small) corner case, which is why I ask about it.

File: pandas/tests/io/test_excel.py [hunk 1]
```diff
@@ -995,11 +995,18 @@ def test_read_excel_nrows_greater_than_nrows_in_file(self, ext):
 
     def test_read_excel_nrows_non_integer_parameter(self, ext):
```
Comment (WillAyd) [inline pandas/tests/io/test_excel.py]: Would be better to parametrize this test for cases like `[-1, 0, '1']` instead of having separate tests

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/tests/io/parser/common.py [hunk 2]
```diff
@@ -375,6 +375,9 @@ def test_read_nrows(self):
         with tm.assert_raises_regex(ValueError, msg):
             self.read_csv(StringIO(self.data1), nrows=-1)
 
+        with tm.assert_raises_regex(ValueError, msg):
```
Comment (gfyoung) [inline pandas/tests/io/parser/common.py]: Can you reference the issue number in a comment above this line.

**Comment by jschendel** [comment]
I'm not sure about disallowing `nrows=0`.  It's generally useful when you need to dynamically read a subset of columns, where you do a two pass procedure along the lines of:

1. Read in the columns only, e.g. `all_cols = pd.read_csv('file.csv', nrows=0).columns`
2. Dynamically filter the columns, e.g. `good_cols = [c for c in all_cols if c.startswith('foo')]`
3. Read in only the desired columns, e.g. `df = pd.read_csv('file.csv', usecols=good_cols)`

I doubt you lose much performance by using `nrows=1` instead, so it's not the end of the world if this does get removed.  However, there are some established StackOverflow answers with `nrows=0` similar to my procedure above, so I imagine there are some users that have `nrows=0` in their codebase.  If this is going to be removed, it definitely seems like it should be deprecated first.

**Comment by gfyoung** [comment]
@jschendel : Good to know.  In the case of your example, couldn't you just read the first line of the file using the builtin `open` ?  Don't really need `pandas` for that. 🤷‍♂️

**Comment by jschendel** [comment]
Yeah, I'm not saying that it's the _right_ solution, just that it's a solution people are likely using.  Certainly more concise than using `open` if you method chain to do the filtering.  I wouldn't complain if this was removed, but doing so without deprecating seems likely to break things for users.

**Comment by gfyoung** [comment]
> I wouldn't complain if this was removed, but doing so without deprecating seems likely to break things for users.

**Review by jreback — CHANGES_REQUESTED** [review]
nrows=0 is valid
the combinations of options is invalid

pr needs to limit scope

**Comment by gfyoung** [comment]
> nrows=0 is valid
the combinations of options is invalid

@jreback : `nrows=0` has limited use and was a clean way to patch the corner case (though it seems like @jschendel has a case to keep it).  In any case, should we just disallow that combination of inputs then?

**Comment by cgopalan** [comment]
@jreback @gfyoung i too agree that just the value of `low_memory` should not render the combination of values invalid. Because of this corner case, this is actually exhibiting weird behavior on an unrelated set of conditions. Better to suppress this corner case by removing/deprecating `nrows=0`.
Is there a way to deprecate just a value for an arg? I would assume we want to keep `nrows` as an argument.

**Comment by jreback** [comment]
> @jreback @gfyoung i too agree that just the value of low_memory should not render the combination of values invalid. Because of this corner case, this is actually exhibiting weird behavior on an unrelated set of conditions. Better to suppress this corner case by removing/deprecating nrows=0.
Is there a way to deprecate just a value for an arg? I would assume we want to keep nrows as an argument.

no reason to remove or even deprecate ``nrows=0``, you simply need to addees for this particular combination of options

**Comment by cgopalan** [comment]
@jreback so `read_csv(StringIO(data), low_memory=False, index_col=0, nrows=0)` is allowed but `read_csv(StringIO(data), low_memory=True, index_col=0, nrows=0)` ?

File: pandas/tests/io/test_excel.py (hunk 1 above)
Comment (cgopalan) [reply pandas/tests/io/test_excel.py]: @WillAyd sure, I will make these changes as soon as we decide if we are keeping this PR :)

**Comment by jreback** [comment]
yeah i thought that was fine

**Comment by gfyoung** [comment]
@jreback : 🤦‍♂️ well, I guess we're heading back there then. 😄

**Comment by cgopalan** [comment]
@gfyoung should i delete this branch?

**Comment by gfyoung** [comment]
@cgopalan : Sure thing.

**Pull request closed without merging**